### PR TITLE
refactor word lowering

### PR DESCRIPTION
**Pull request by jackzhenguo**

### CODE REVIEW

File: S21/jackzhenguo/refactorme.py [hunk 1]
```diff
@@ -1,84 +1,112 @@
 import unicodedata
 
+
 class Word:
+    def __init__(self, word, bcp_code):
+        self.wc = word
+        self.lang = bcp_code
+
+
+class Strategy(object):
```
Comment (kscanne) [inline S21/jackzhenguo/refactorme.py:10]: I'd suggest "LowercaseStrategy" for this.

Comment (kscanne) [reply S21/jackzhenguo/refactorme.py:10]: You're tying this to the lowercasing application by including the Word as an attribute. If you want to preserve the generic strategy interface, you'd need that as a parent of LowercaseStrategy, with no attribute "word".

File: S21/jackzhenguo/refactorme.py [hunk 2]
```diff
@@ -1,84 +1,112 @@
...
+        """
+        return self.word.wc
+
+
+class StrategyVowel(Strategy):
+    """
+    strategy of processing to vowel cases
+    """
+    def process(self) -> str:
+        vowels = 'AEIOU\u00c1\u00c9\u00cd\u00d3\u00da'
```
Comment (kscanne) [inline S21/jackzhenguo/refactorme.py:43]: Magic constants!

File: S21/jackzhenguo/refactorme.py [hunk 3]
```diff
@@ -1,84 +1,112 @@
...
+        temp = self.word.wc
+        if len(temp) > 1 and (self.word.wc[0] in ['t', 'n']) and \
+                unicodedata.normalize('NFC', self.word.wc)[1] in vowels:
+            temp = self.word.wc[0] + '-' + temp[1:]
+        return temp.lower()
+
+
+class StrategyCharReplace(Strategy):
```
Comment (kscanne) [inline S21/jackzhenguo/refactorme.py:51]: Might tie the name to the specific case here (Turkish script)

File: S21/jackzhenguo/refactorme.py [hunk 4]
```diff
@@ -1,84 +1,112 @@
...
+        """
+        strategy of replacing single char
+        :return:
+        """
+        temp = self.word.wc
+        temp = temp.replace('\u0049', '\u0131')
+        return temp.lower()
+
+
+class StrategyTrail(Strategy):
```
Comment (kscanne) [inline S21/jackzhenguo/refactorme.py:62]: Same; this is specifically Greek.  You could imagine something more abstract that allows an implementer to specify the trailing character, but that's not what you have here.

File: S21/jackzhenguo/refactorme.py [hunk 5]
```diff
@@ -1,84 +1,112 @@
...
+        temp = self.word.wc
+        if temp[-1] == '\u03a3':
+            temp = temp[:-1] + '\u03c2'
+        return temp.lower()
+
+
+class WordProcess(object):
+    def __init__(self, word: Word):
+        self.word = word
+        self.strategy_dict = {'en': StrategyBase, 'zh': StrategyKeep, 'ja': StrategyKeep,
```
Comment (kscanne) [inline S21/jackzhenguo/refactorme.py:76]: Lovely.

File: S21/jackzhenguo/refactorme.py [hunk 6]
```diff
@@ -1,84 +1,112 @@
...
+
+    def to_lower(self):
+        language = self.convert_lang()
+        if language in self.strategy_dict:
+            specific_strategy = self.strategy_dict[language]
+        else:
+            specific_strategy = StrategyBase  # default strategy
+        return specific_strategy(self.word).process()
+
+    def convert_lang(self) -> str:
```
Comment (kscanne) [inline S21/jackzhenguo/refactorme.py:88]: I'd like this moved to it's own class representing BCP-47 strings.

**Review by kscanne — COMMENTED** [review]
Very nice implementation.

File: S21/jackzhenguo/refactorme.py (hunk 6 above)
Comment (jackzhenguo) [reply S21/jackzhenguo/refactorme.py:88]: Thank you for giving me such good suggestions. Actually, we have done another version yesterday from our team that Willie pulled request.

**Pull request closed without merging**